### DISCUSSION TITLE
Lots of "error" log level messsages (but working fine): false-positives/normal?

### DISCUSSION THREAD

**Discussion by pn-santos** [concourse/concourse#7217] [Help & Support] [answered]
Hi, in our concourse setup we notice quite a few "types/classes" of `level: error` messages that repeat over and over. I am a bit unsure if they:

- signal an actual _error_ (i.e. a bug in concourse or some issue with config/network or db state that operators should look into fixing or improving) 
- they are _errors_ but part of _normal_ operation (e.g. something failing due to a network timeout because, well network was slow at the time)
- aren't really errors (and likely should either not be logged or logged at another level) 

We're on version **7.3.2** - There's nothing apparently wrong that we can see that users would notice (at least nobody is shouting about it), so it's not like something seems fundamentally broken.

I'll list some of the ones that look more "scary/unexpected" ones in the hopes someone might shed a light on what they mean (and whether ops should be worried about them). These happen on a daily basis, some of them multiple times a day:

 ```
pq: deadlock detected
```
in messages:
- `atc.collector_checks.tick.component-failed`
- `atc.collector_checks.tick.check-collector.failed-to-delete-completed-checks`
- `atc.collector_resource_cache_uses.tick.resource-cache-use-collector.failed-to-clean-finished-build-uses`

---

```
pq: new row for relation \"workers\" violates check constraint \"workers_active_tasks_check\"
```
in messages: 
- `atc.tracker.notify.run.task-step.failed-to-decrease-active-tasks`

---

```
resource type '...' has no version
```
(happens with several resource types, e.g. with https://github.com/telia-oss/github-pr-resource but with none of the included resource types that come with concourse - we get _thousands_ of these logs every day)
in messages:
- `atc.scanner.tick.failed-to-create-check`

---

Then there's quite a few logs around network connections which I guess could just be the way it goes (no network is 100% reliable) but just in case they might point to some underlying issue with our setup or with concourse itself I'll post them anyway:

```
Put \"/volumes/bbc054e1-9cdc-460e-6b5a-aa2f4de13502/stream-in?path=.\": http: unexpected EOF reading trailer
```
in messages:
- `atc.tracker.notify.run.put-step.failed-to-create-volume-mounts-for-container`:

---

```
Get \"http://ccm1:33827/volumes\": dial tcp 192.168.39.49:33827: connect: connection refused
```
in messages;
- `tsa.connection.channel.command.heartbeat.failed-to-list-volumes`

---

```
ssh: unexpected packet in response to channel open: \u003cnil\u003e
```
in messages:
- `tsa.connection.request.forward-conn.failed-to-open-channel`

---

```
accept tcp [::]:33827: use of closed network connection
```
in messages:
- `tsa.connection.request.failed-to-accept`

**Comment by taylorsilva** [comment]
For:
```
pq: new row for relation \"workers\" violates check constraint \"workers_active_tasks_check\"
```
in messages: `atc.tracker.notify.run.task-step.failed-to-decrease-active-tasks`

This error happens when `active_tasks` is zero for a worker and we try to decrease the value past zero since there's a constraint on the column that the value has to be `>= 0`. Might be a symptom related to #7085

Question is, how is Concourse decreasing by more tasks than it initially incremented by? Definitely a bug.

**Reply by taylorsilva** [reply]
What container placement strategies are you using?

**Reply by pn-santos** [reply]
`limit-active-containers → limit-active-volumes → limit-active-tasks → fewest-build-containers`

**Reply by taylorsilva** [reply]
I tried to reproduce this by creating a Concourse cluster with those placement strategies. I only limited active tasks to `2`, everything else I left as unlimited. I set a pipeline that had a "hello-world" task and triggered 100 builds. I grep'd the logs for anything with `pq` but got zero hits. I'm not sure what else to try to get this bug to reproduce locally. 

Looking at the code, I didn't notice any area where we may have meant to increment active tasks but don't. I'm guessing this is some rare race condition. Maybe it's more obvious in large environments?

**Reply by pn-santos** [reply]
What would you recommend? we've tried bringing every ATC and worker down (and reset the workers) and bring them back up but that didn't solve it. Is there any state associated with tasks in the DB that we can clear while the system is down to "reset" the active tasks tracking? or is it all kept in-memory?

**Reply by taylorsilva** [reply]
The `workers` table has an `active_tasks` column. You can set that to zero after bringing down the cluster to "reset" everything.

**Comment by taylorsilva** [comment, marked as answer]
```
resource type '...' has no version
```

Lidar creates checks for resource-types first to avoid this exact issue. There was a refactor to this code recently but it didn't change that behaviour: https://github.com/concourse/concourse/pull/6923

This does eventually resolve itself, annoying that you're getting this error though. I wonder why you're getting it so often? I would expect this to occur once for a given resource type and then never again once the version for the parent resource type is found. Is something clearing out the version(s) for the resource type?

**Reply by pn-santos** [reply]
> Is something clearing out the version(s) for the resource type?

Not that I am aware of. I did try doing `fly -t <target> crt` on a random pipeline resource type that shows up in the logs with this message but it worked fine... 

Not sure if it could be a specific pipeline causing this but we do get a LOT of messages in the last minute there was 44 messages. The messages seem to happen in "spikes" (11 messages exactly all in the same millisecond). They seem to repeat every ~10s sometimes a 20s interval. But pretty constant and always occurring in "bursts" of exactly 11 messages.

**Reply by taylorsilva** [reply]
Lidar does its scan every 10s. That sounds like it matches up with that. Maybe you have a bad pipeline somewhere with resource-types that are incorrectly configured, trying to pull an image that doesn't exist maybe?

**Reply by pn-santos** [reply]
Makes sense, I'll try and dig for it see if I can find the bad pipeline(s) :+1:

**Reply by pn-santos** [reply]
So turns out the issue wasn't with the actual `resource_types` at all but with resource _of_ those types.

Some resources (of those types) had bad access tokens or credential values (which are fetched from vault). The resources in question didn't have any versions and when trying to do the first check were getting a "unauthorized" (401) responses.

Once we fixed the access token values, the `no version` error messages went away. 

It's a bit weird though that a _resource_ having no versions would cause a message about a resource _type_ having no versions... from the outside I'd say the log message could be worded a bit better (assuming at the place it's logged one can distinguish between `resource_type` checks and `resource` checks?). 

If it had been something like:

> resource **_of_** type '...' has no version

would have helped me look in the right place ^_^ (or even maybe saying the resource name? not sure if it is available where the message is logged)

**Reply by taylorsilva** [reply]
Hmm I wonder if we have something backwards in the code. It reads fine to me and you should see this error when a type of a resource has no versions. But it was the other way for you? The type was fine but the resource was failing to produce versions because of a config error? I'm surprised that wasn't surfaced in the output of theresource's `check`.
https://github.com/concourse/concourse/blob/459b727875456577dbe94422528988f9dc62f790/atc/db/check_factory.go#L92-L96

**Reply by pn-santos** [reply]
Humm I guess it could all have been a coincidence and it was something else that fixed it :man_shrugging:. If I have time later on I can try and reproduce it (should be easy to test the scenario I _think_ caused it). 

But yeah, I look at that piece of code and couldn't see an obvious mistake (that's why I'm not very confident about the root cause, could all have been a coincidence that we fixed the resources at the same time someone somewhere fixed a pipeline).

**Reply by pn-santos** [reply]
@taylorsilva  I finally had some time to try and reproduce this and it's fairly easy tbh. Here's the setup:
```yaml
resource_types:
- name: github-pr
  type: registry-image
  source:
    repository: teliaoss/github-pr-resource
    tag: v0.23.0

resources:
- name: pr.git
  type: github-pr
  icon: source-pull
  source:
    access_token: ((github-repo-access-token))
    repository: some-org/some-repo
```

In my setup `((github-repo-access-token))` doesn't exist in vault. On the first resource check concourse UI shows this:
![unauthorized](https://user-images.githubusercontent.com/23340138/125645372-18357323-17dc-447f-ae0c-610ae3a4f08d.png)

the logs shows this:
![error](https://user-images.githubusercontent.com/23340138/125645437-5fe946c2-f0ad-4074-a4e4-1639318a47dc.png)

**Reply by taylorsilva** [reply]
I got that message in my logs to appear once using your example. It happened immediately after unpausing the pipeline. I'm realizing now that this makes sense that you get the error initially. All lidar does is "schedule a check". It takes time for the other atc components to pick up and DO the work of running the check. So there is a period of time where there would be no versions for the base resource-type, so lidar would fail to make checks for the resources BASED on that resource-type. But once there is a version you'll stop seeing this error in your logs. 

Still not sure why you keep seeing this message so many times in your logs. Unless someone is recreating their pipeline constantly or changing the `source` of their resource-types, you shouldn't see the error that often.

**Reply by pn-santos** [reply]
Aaannnd you are right, it does only log that message once and subsequent checks don't trigger it (I should have done my due diligence and waited to see it repeat)... 

It must have been some pipeline that had bad resource type configs. I did find 3 repos that had typos on `type` (they used `type: regitsry-image` instead of `registry-image`) but they were all on the same resource type so it didn't explain why more than one resource type was showing up in the logs with the "no version" message, but maybe there were more of the bad resource types? :shrug:  error is gone now so everything is fine now.

Thanks for sticking in and help out trying to get to the bottom of it :+1:

**Comment by taylorsilva** [comment]
The networking messages simply read like networking flakes to me. There's not much interesting code-wise there, we make the network call and check if it succeeds 🤷
